### DISCUSSION TITLE
part parameters and special characters

### DISCUSSION THREAD

**Discussion by T0jan** [inventree/InvenTree#5140] [General]
Hi,

until 0.12.0 there was no filtering in what could be written into a parameter value and in general I like the idea to limit it. However I think the current implementation is to harsh as it also rejects special characters. Like this range definitions or the standard plus minus sign as seen in the following example are no longer viable unless you don't define a standard unit which in my view contradicts itself:

![image](https://github.com/inventree/InvenTree/assets/22519396/9c9568bf-f3c5-4b25-b6d0-8cb486ce5e3d)

It also significantly complicates automatic data import from distributor APIs like digikey where the parameters usually come with special characters, where before you could use the value as it came you now have to modify and remove characters to be able to upload a parameter via the InvenTree API.

leaving fields empty, e.g. with a simple `-` as indicator that the parameter is not known for this part now is only possible by entering the unit instead into the fields which is a lot less clean looking.

So I would strongly suggest to rethink this filtering approach here in favor of usability.

**Comment by matmair** [comment]
@T0jan have you tried not setting a unit for the field? Having it both ways on one parameter will be hard/impossible.

**Comment by SchrodingersGat** [comment]
@T0jan the move to physical units does impose some restrictions, which might require some thought to work around. 

If units are specified for a parameter, the provided value must correspond to the specified unit. The pint library does not support +/- characters, unfortunately.

**Comment by T0jan** [comment]
@matmair @SchrodingersGat I mean sure one could leave the unit field now blank for all parameters to still allow automatic import but that makes the unit field in itself pretty much obsolete.

**Reply by SchrodingersGat** [reply]
For parts and supplier parts, the "units" field *must* be valid - so that the part quantity is converted correctly based on supplier part stock, incoming orders, etc.

However for part parameters, there is probably nothing *terrible* about allowing invalid values to be input. e.g. "-50 ~ 125" is not a *valid* temperature value - here "valid" means it can be converted to a native value which can be compared against other values. However, the only downside to allowing it to be input is that the numerical sorting won't work properly.

I'd be open to loosening up the restrictions here for part parameters, so that the "units" must be a valid unit type, and if the user inputs a strange value, it is allowed, but won't play nice with natural sorting.

**Reply by T0jan** [reply]
yeah I think there is no case where an exception for part "units" would make sense as it always should be a defined number. allowing "strange" values for parameters only sounds like a good idea.

**Reply by SchrodingersGat** [reply]
@T0jan see https://github.com/inventree/InvenTree/pull/5160

**Reply by T0jan** [reply]
thanks for the fast work as always :)

**Reply by SchrodingersGat** [reply]
@T0jan [this PR](https://github.com/inventree/InvenTree/pull/5175) makes this a user configurable option

**Reply by T0jan** [reply]
@SchrodingersGat seems to work except for one minor thing: if `-` is used as placeholder in case a parameter value is not known the server throws an AttributeError:

```
requests.exceptions.HTTPError: {'detail': 'Error occurred during API request', 'url': 'https://inventree-test/api/part/parameter/', 'method': 'POST', 'status_code': 500, 'body': '{"error":"AttributeError","error_class":"<class \'AttributeError\'>","detail":"\'NoneType\' object has no attribute \'evaluate\'","path":"/api/part/parameter/","status_code":500}', 'headers': {'AUTHORIZATION': 'Token ...'}, 'params': {'format': 'json'}, 'data': {'part': 439, 'template': 15, 'data': '-'}}
```

However I think it is not critical, one could just use a different placeholder like an underscore.

**Reply by SchrodingersGat** [reply]
@T0jan thanks for pointing that out, that is a separate bug - https://github.com/inventree/InvenTree/pull/5183
